pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/SnapshotSegmentAbortedTxnProcessorImpl.java

```diff
@@ -0,0 +1,774 @@
+                    });
+            res.exceptionally(e -> {
+                //Just log the error, and the processor will try to take snapshot again when the transactionBuffer
+                //append aborted txn next time.
+                log.error("Failed to take snapshot segment [{}] at maxReadPosition [{}] "
+                                + "for the topic [{}], and the size of the segment is [{}]",
+                        this.sequenceID, abortedMarkerPersistentPosition, topic.getName(),
+                        sealedAbortedTxnIdSegment.size(), e);
+                return null;
+            });
+            return res;
+        }
+
+        private CompletableFuture<Void> writeSnapshotSegmentAsync(LinkedList<TxnID> segment,
+                                                                  PositionImpl abortedMarkerPersistentPosition) {
+            TransactionBufferSnapshotSegment transactionBufferSnapshotSegment = new TransactionBufferSnapshotSegment();
+            transactionBufferSnapshotSegment.setAborts(convertTypeToTxnIDData(segment));
+            transactionBufferSnapshotSegment.setTopicName(this.topic.getName());
+            transactionBufferSnapshotSegment.setPersistentPositionEntryId(abortedMarkerPersistentPosition.getEntryId());
+            transactionBufferSnapshotSegment.setPersistentPositionLedgerId(
+                    abortedMarkerPersistentPosition.getLedgerId());
+
+            return snapshotSegmentsWriterFuture.thenCompose(segmentWriter -> {
+                transactionBufferSnapshotSegment.setSequenceId(this.sequenceID.get());
+                return segmentWriter.writeAsync(buildKey(this.sequenceID.get()), transactionBufferSnapshotSegment);
+            }).thenCompose((messageId) -> {
+                PositionImpl maxReadPosition = topic.getMaxReadPosition();
+                //Build index for this segment
+                TransactionBufferSnapshotIndex index = new TransactionBufferSnapshotIndex();
+                index.setSequenceID(transactionBufferSnapshotSegment.getSequenceId());
+                index.setAbortedMarkLedgerID(abortedMarkerPersistentPosition.getLedgerId());
+                index.setAbortedMarkEntryID(abortedMarkerPersistentPosition.getEntryId());
+                index.setSegmentLedgerID(((MessageIdImpl) messageId).getLedgerId());
+                index.setSegmentEntryID(((MessageIdImpl) messageId).getEntryId());
+
+                indexes.put(abortedMarkerPersistentPosition, index);
+                //update snapshot segment index.
+                //If the index can not be written successfully, the snapshot segment wil be overwritten
+                // when the processor writes snapshot segment next time.
+                return updateSnapshotIndex(new TransactionBufferSnapshotIndexesMetadata(
+                                maxReadPosition.getLedgerId(), maxReadPosition.getEntryId(),
+                                convertTypeToTxnIDData(unsealedTxnIds)));
```

Comment:
Is there a scenario like this? 

- The field `indexes` is `{aborts=[], maxReadPosition=-1:-1, segments=[] }` 
- append task `write_segment_1` to the queue,  the max read position is `3:3` now.
- append task `write_segment_2` to the queue,  the max read position is `5:5` now.
- execute task `write_segment_1`
  - update indexes to `[{aborts=[], maxReadPosition=5:5}, segments=[s1]]` 
- broker crash

an error occurs: then segment 2 was lost, but the maxReadPosition of the topic is `5:5`.

Reply:
Fixed, I make the maxReadPosition is passed at appending write segment task.

Reply:
| time | thread-`append task-write-segment` | thread-`execute task which in the queue` |
| --- | --- | --- |
| 1 | | write `segment-1` |
| 2 | | check: only has one task in the queue(`true`) |
| 3 | append new task `write-segment-2` | |
| 4 | clean unsealed list | |
| 5 | | update indexes |

if the broker crash after step 5: after reloading this topic, an error occurs: the maxReadPosition is the newest `maxReadPosition`, but segment 2 was lost.

Reply:
I have thought about this. So I first set maxReadPosition and unsealed Aborts as local variables, and then make a judgment.

```
 PositionImpl maxReadPosition = topic.getMaxReadPosition();
            List<TxnIDData> aborts = convertTypeToTxnIDData(unsealedTxnIds);
            if (taskQueue.size() != 1) {
```

Reply:
| time | thread-`TB.abortTxn` | thread-`execute task which in the queue` |
| --- | --- | --- |
| 1 | | write `segment-1` |
| 2 | update `maxReadPosition` of the topic | | 
| 3 | | read `maxReadPosition` of the topic | 
| 4 | | read `unsealedAborts` | 
| 5 | put aborts to `unsealedAborts` | | 
| 6 | | update indexes | 

if the broker crash after step 6: after reloading this topic, an error occurs, the `maxReadPosition` is larger than `unsealedAborts`

Reply:
Good catch! Fixed.